### PR TITLE
Native AOT runtime changes for multiple package conflict

### LINKED ISSUE #26268: Make it possible to use newer ILCompiler packages than what SDK comes with
### Describe the bug
Publishing app with NativeAOT results in broken binary.

### To Reproduce
- Create C# hello world console app and edit project file to be like this
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net7.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <PublishAot>true</PublishAot>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.DotNet.ILCompiler" Version="7.0.0-preview.6.22325.1" />
  </ItemGroup>

</Project>
```
Presence of both **PublishAot** and **PackageReference** to Microsoft.DotNet.ILCompiler is important.
- Publish app (win-x64) and then run it
- You will get exception like this
```
Unhandled Exception: System.InvalidOperationException: Operation is not valid due to the current state of the object.
   at Internal.Reflection.Core.Execution.ReflectionCoreExecution.InitializeExecutionDomain(ReflectionDomainSetup, ExecutionEnvironment) + 0x94
   at Internal.Reflection.Execution.ReflectionExecution.Initialize() + 0x35
   at ConsoleApp5!<BaseAddress>+0x1f0d6a
```

### Further technical details
- Tested with .NET 7 SDK Preview 5 and also .NET 7 SDK daily build
- Here is why it doesn't work: build task that generates AppName.ilc.rsp file for ilc compiler adds duplicate **--initassembly** arguments
```
--initassembly:System.Private.CoreLib
--initassembly:System.Private.StackTraceMetadata
--initassembly:System.Private.TypeLoader
--initassembly:System.Private.Reflection.Execution
--initassembly:System.Private.CoreLib
--initassembly:System.Private.StackTraceMetadata
--initassembly:System.Private.TypeLoader
--initassembly:System.Private.Reflection.Execution
```

### PR DESCRIPTION
**Pull request by LakshanF**
Fixes https://github.com/dotnet/sdk/issues/26268 where currently adding a ILCompiler package reference creates an application that doesn't work. We want to support this scenario where we expect developers to get the latest ILCompiler package (rather than use the SDK one, which can be a little older) when publishing, for example to pick a fix that is not yet present in the SDK. 

Some notes

- The fix is going to be in the SDK (https://github.com/dotnet/sdk/pull/26652) but requires these changes. The SDK PR will be in draft form until these changes are merged and flow into SDK. It is important to validate that the fix works due to the delay in syncing
- There are 2 packages that are affected: `microsoft.dotnet.ilcompiler `deals primarily with setting up properties and targets and is handled similar to the way the linker manages conflicts with package references (see https://github.com/dotnet/sdk/pull/25993 and https://github.com/dotnet/linker/pull/2837)

> 1. SDK-ONLY: In the SDK, import the `ILCompiler` package, `Sdk.props` from `Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.props`

> 2. SDK-ONLY: `ILCompiler's Sdk.props` will only import `Microsoft.DotNet.ILCompiler.props` from the `ILCompiler` package build directory if a sentinel, `ILCompilerTargetsPath`, is not set. The SDK will also set a flag, `AotRuntimePackageLoadedViaSDK`, to indicate that the SDK has loaded the `ILCompiler `package if the sentinel is not previously set.

> 3. Both SDK and Nuget: `Microsoft.DotNet.ILCompiler.props` contains a sentinel, `ILCompilerTargetsPath`, that sets the path to `Microsoft.DotNet.ILCompiler.targets`. We assume that if there is an explicit package reference to the `ILCompiler` package, Nuget will set this sentinel and SDK will honor that. If there is no explicit package reference, then the SDK will set the sentinel

> 4. There will be only one path, the sentinel path, that will be used to set necessary properties and targets when there are multiple references to the `ILCompiler` package with this approach. We want to use the explicit package reference if the developer expresses the intent. `Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets` will load this sentinel and will honor the explicit package reference path if one is present.

- The platform specific package (for example, `runtime.win-x64.microsoft.dotnet.ilcompiler` for the windows package) that contains ilc tool and runtime assets. The [package reference path](https://github.com/dotnet/runtime/blob/main/src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.targets#L43-L63) is now set based on a flag that gets sets if the SDK package is the only one used. Otherwise, the Nuget package one that comes from the explicit package reference in the app will be picked.
- The publish AOT mechanism has evolved over the time and needs to be cleaned up. Issue #72415 tracks this and should be done as a high priority since following the current code is painful

> 1. The original code was added to only support explicit package references

> 2. SDK support was added for `PublishAot` with backward compatibility support for the explicit package reference without the SDK path (for a short period)

> 3. This feature adds support to adding explicit package reference with SDK `PublishAot`.  We are not planning to support the explicit package reference without the SDK path.

### CODE REVIEW

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.targets
Comment (agocke) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.targets:52]: Can we somehow reuse this logic for the NuGet package as well? i.e., have the NuGet package set `ResolvedILCompilerPack` and use the SDK logic for restoring the runtime pack?

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.targets
Comment (LakshanF) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.targets:45]: To the best of my knowledge, AotRuntimePackageLoadedViaSDK, is intended to indicate `ILCompiler `package loaded mechanism and is a new flag. `IlcCalledViaPackage `seems to used to control the native targets and is used for the `ILCompiler `runtime package

Comment (LakshanF) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.targets:45]: Created an issue to track this, #72415

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.targets
Comment (LakshanF) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.targets:52]: Created an issue to track this, #72415

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.targets
Comment (MichalStrehovsky) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.targets:45]: IlcCalledViaPackage was introduced here: https://github.com/dotnet/corert/pull/5123 - all the context we have left is in what's in the pull request/comments.

It was working around some issue around two situations:
* The props/targets are included from the NuGet package (IlcCalledViaPackage == true) and
* The props/targets are included directly (like we do in this repo when we build tests, or crossgen2)

There were things happening differently in those two scenarios. I'm not aware of us doing anything to make that not needed, but I also don't know much about what the original problem was.

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.targets
Comment (agocke) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.DotNet.ILCompiler.targets:52]: That would be my hope, but I've noticed that restore is weird and delicate, so it may not work how I would expect.

File: src/installer/pkg/projects/Microsoft.DotNet.ILCompiler/sdk/Sdk.props
Comment (LakshanF) [reply src/installer/pkg/projects/Microsoft.DotNet.ILCompiler/sdk/Sdk.props:20]: I wasn't sure if there is any way that the SDK path can be activated before the explicit ILCompiler package reference can set the sentinel since there are non-PublishAOT calls in the [SDK](https://github.com/dotnet/sdk/pull/26652/files#diff-cde7a8c0e370b40890bd6539212594fdbb8688a55f9f40524d535c3972dd54a6R156) that might set the sentinel. 

So there are two concerns here,

- If explicit `ILCompiler` package references always set the sentinel first, then its ok to remove the check here
- There needs to be a guard for non-PublishAot SDK code paths to not import aot targets since [Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.props](https://github.com/dotnet/sdk/pull/26652/files#diff-cde7a8c0e370b40890bd6539212594fdbb8688a55f9f40524d535c3972dd54a6) and [Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets](https://github.com/dotnet/sdk/pull/26652/files#diff-b8dad9da3662ff7878709ecc6bba9f3249c3dd1471b3e67ae13ee6d009604040) might get invoked in non-PublishAot scenarios. I also added a check in the targets file before importing the native AOT target file and if the above concern is not valid, we can remove it from the props file

**Review by sbomer — APPROVED** [review]
Functionally LGTM. My personal preference would be to clean up the old properties here, but it also seems fine to do that in a follow-up if you want to unblock this change. Thank you!

**Comment by kasperk81** [comment]
this is blocking sdk flow https://github.com/dotnet/sdk/pull/26715

and is it really needed by even 1% of the users? it is disconnected from everything else integrated into the sdk. crossgen integration in sdk is shipped without this hook and no one has ever asked or complained about it. we  wait for next release of .net for all the new features and improvements together, if someone need to try out the latest dev version once in a while. they download the daily build which takes no more than a few seconds to download than the runtime pack..

**Comment by jkotas** [comment]
@kasperk81 Both linker and Roslyn have similar hooks. Yes, we need this hook to be successful with customers that we are closely working with.

crossgen does not have this hook since it is tightly coupled with the runtime pack and so it does not make sense to have a separate hook for it.

> they download the daily build which takes no more than a few seconds to download than the runtime pack.

The SDK version is often controlled in customer build environments. Requiring them to update the SDK to pickup or validate a bug-fix is a non-starter.

**Pull request merged**
